### PR TITLE
(cloud-merge) Support to abort txn when coordinate be restart and do schema change

### PR DESCRIPTION
**Pull request by Lchangliang**
When do schema change, it will check whether transactions of the given database which txnId is less than 'watershedTxnId' are finished. If is not, schema change will be waiting.
When the coordinator restart (whatever fe/be), the txn belong the cooridinator will be hang until timeout. So we need to abort it in time.
There are two optimizations.
1. Check BE's lastStartTime when get it heart beat. If current lastStartTime is large than lastStartTime in fe memory. Abort all the hang txns belong  this BE.
2. Check conflict txns when do schema change. If txns is failed (Maybe coordinator be/fe restart), abort it directly.

### CODE REVIEW

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run cloud_p0

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java [hunk 1]
```diff
@@ -442,6 +444,54 @@ public boolean existCommittedTxns(Long dbId, Long tableId, Long partitionId) {
...
+        TxnCoordinator coordinator = txn.getCoordinator();
+        if (coordinator.sourceType == TransactionState.TxnSourceType.FE) {
+            List<Frontend> frontends = Env.getCurrentEnv().getFrontends(null);
+            for (Frontend fe : frontends) {
+                if (fe.getHost().equals(coordinator.ip) && fe.getLastStartupTime() > coordinator.startTime) {
+                    return true;
+                }
+            }
+        } else if (coordinator.sourceType == TransactionState.TxnSourceType.BE) {
+            Backend be = Env.getCurrentSystemInfo().getBackend(coordinator.id);
```
Comment (luwei16) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:457]: Consider the case where be is offline

File: fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeJobV2.java [hunk 2]
```diff
@@ -792,10 +796,20 @@ private void cancelInternal() {
...
+        List<TransactionState> unFinishedTxns = Env.getCurrentGlobalTransactionMgr().getUnFinishedPreviousLoad(
                 watershedTxnId, dbId, Lists.newArrayList(tableId));
+        if (Config.enable_abort_txn_by_checking_conflict_txn) {
+            List<TransactionState> failedTxns = GlobalTransactionMgr.checkFailedTxns(unFinishedTxns);
+            for (TransactionState txn : failedTxns) {
+                Env.getCurrentGlobalTransactionMgr()
+                        .abortTransaction(txn.getDbId(), txn.getTransactionId(), "Cancel by schema change");
+            }
+        }
+        return unFinishedTxns.isEmpty();
```
[the remark this answers is not in the document]
Comment (Lchangliang) [reply fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeJobV2.java:812]: No, if unFinishedTxns is not empty, it need to abort the txns in current term and doublc check in next term.

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run external

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 3]
```diff
@@ -2553,6 +2568,100 @@ void MetaServiceImpl::abort_sub_txn(::google::protobuf::RpcController* controlle
...
+                ss << "malformed txn running info";
+                msg = ss.str();
+                ss << " key=" << hex(k);
+                LOG(WARNING) << ss.str();
+                return;
+            }
+            const auto& coordinate = info_pb.coordinator();
+            if (info_pb.status() == TxnStatusPB::TXN_STATUS_PREPARED &&
+                coordinate.sourcetype() == TXN_SOURCE_TYPE_BE && coordinate.id() == request->id() &&
+                coordinate.ip() == request->ip() &&
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:2633]: what if there are multiple BEs on the same host?

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 4]
```diff
@@ -2553,6 +2568,100 @@ void MetaServiceImpl::abort_sub_txn(::google::protobuf::RpcController* controlle
...
+        err = txn->get(begin_info_key, end_info_key, &it, true);
+        if (err != TxnErrorCode::TXN_OK) {
+            code = cast_as<ErrCategory::READ>(err);
+            ss << "failed to get txn info. err=" << err;
+            msg = ss.str();
+            LOG(WARNING) << msg;
+            return;
+        }
+
+        while (it->has_next()) {
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:2617]: Is it possible that this loop take more than 5 seconds which will lead to "TxnTimeout Error"?

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 5]
```diff
@@ -2640,10 +2750,43 @@ void MetaServiceImpl::check_txn_conflict(::google::protobuf::RpcController* cont
...
+                        // decode version key error means this is something wrong,
+                        // we can not continue this txn
+                        LOG(WARNING) << "failed to decode key, ret=" << ret << " key=" << hex(k);
+                    } else {
+                        DCHECK(out.size() == 5) << " key=" << hex(k) << " " << out.size();
+                        const std::string& decode_instance_id = std::get<1>(std::get<0>(out[1]));
+                        int64_t db_id = std::get<0>(std::get<0>(out[3]));
+                        int64_t txn_id = std::get<0>(std::get<0>(out[4]));
+                        std::string conflict_txn_info_key =
+                                txn_info_key({decode_instance_id, db_id, txn_id});
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:2768]: make this part a separate function which should be called `get_txn_info_key_from_txn_running_key()`

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: cloud/src/meta-service/meta_service_txn.cpp (hunk 3 above)
Comment (Lchangliang) [reply cloud/src/meta-service/meta_service_txn.cpp:2633]: "coordinate.id() == request->id()"

File: cloud/src/meta-service/meta_service_txn.cpp (hunk 4 above)
Comment (Lchangliang) [reply cloud/src/meta-service/meta_service_txn.cpp:2617]: If is timeout, it means there are too many old unfinished txns need to abort. We can abort they also when check conflict txns.

**Comment by Lchangliang** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java [hunk 6]
```diff
@@ -442,6 +444,56 @@ public boolean existCommittedTxns(Long dbId, Long tableId, Long partitionId) {
...
+    public static List<TransactionState> checkFailedTxns(List<TransactionState> conflictTxns) {
+        List<TransactionState> failedTxns = new ArrayList<>();
+        for (TransactionState txn : conflictTxns) {
+            boolean failed = false;
+            if (!failed) {
+                failed = checkFailedTxnsByCoordinator(txn);
+            }
+            if (failed) {
+                failedTxns.add(txn);
+            }
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:476]:
```suggestion
            if (checkFailedTxnsByCoordinator(txn)) { // failed
                failedTxns.add(txn);
            }
```

File: fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java [hunk 7]
```diff
@@ -442,6 +444,56 @@ public boolean existCommittedTxns(Long dbId, Long tableId, Long partitionId) {
...
+                }
+            }
+        } else if (coordinator.sourceType == TransactionState.TxnSourceType.BE) {
+            Backend be = Env.getCurrentSystemInfo().getBackend(coordinator.id);
+            if (be.getHost().equals(coordinator.ip) && (be.getLastStartTime() > coordinator.startTime
+                    || (!be.isAlive() && System.currentTimeMillis() - be.getLastUpdateMs()
+                                >= Config.abort_txn_after_lost_heartbeat_time_second * 1000L))) {
+                return true;
+            }
+        }
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:463]: add check for dropped FE/BE(dead)

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 8]
```diff
@@ -2553,6 +2568,119 @@ void MetaServiceImpl::abort_sub_txn(::google::protobuf::RpcController* controlle
...
+            ss << "failed to get txn info. err=" << err;
+            msg = ss.str();
+            LOG(WARNING) << msg;
+            return;
+        }
+
+        while (it->has_next()) {
+            total_iteration_cnt++;
+            auto [k, v] = it->next();
+            LOG(INFO) << "check txn info txn_info_key=" << hex(k);
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:2620]: VLOG

File: cloud/src/meta-service/meta_service_txn.cpp (hunk 4 above)
Comment (gavinchou) [reply cloud/src/meta-service/meta_service_txn.cpp:2617]: create_txn every 3 seconds

**Comment by Lchangliang** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by luwei16 — APPROVED** [review]

**Pull request merged**